floracast/steps/evaluate.py

```diff
@@ -0,0 +1,352 @@
+        </body>
+        </html>
+        """
+        return HTMLString(error_html)
+
+
+@step
+def evaluate(
+    model: object,
+    train_series: TimeSeries,
+    val_series: TimeSeries,
+    horizon: int = 7,
+    metric: str = "smape",
+) -> Tuple[
+    Annotated[float, "evaluation_score"],
+    Annotated[HTMLString, "evaluation_visualization"],
+]:
+    """
+    Evaluate the trained model on validation data.
+
+    Args:
+        model: Trained forecasting model
+        train_series: Training time series
+        val_series: Validation time series
+        horizon: Forecasting horizon
+        metric: Evaluation metric name
+
+    Returns:
+        Evaluation metric score (lower is better for SMAPE)
+    """
+
+    logger.info(f"Evaluating with horizon: {horizon}, metric: {metric}")
+
+    try:
+        # Generate predictions using TFT model
+        # TFT requires the series parameter to generate predictions
+        logger.info(f"Generating predictions for horizon: {horizon}")
+
+        # For TFT models, we need to provide the series parameter
+        if hasattr(model, "predict"):
+            # Generate predictions using iterative multi-step approach for longer horizons
+            # This is better than single-shot prediction for long horizons
+            n_predict = min(
+                len(val_series), 42
+            )  # Cap at 6 weeks for reasonable evaluation
+
+            # Use multiple prediction steps for better long-term accuracy
+            predictions_list = []
+            context_series = train_series
+
+            # Predict in chunks of output_chunk_length (14 days)
+            remaining_steps = n_predict
+            while remaining_steps > 0:
+                chunk_size = min(
+                    14, remaining_steps
+                )  # Model's output_chunk_length
+                chunk_pred = model.predict(n=chunk_size, series=context_series)
+                predictions_list.append(chunk_pred)
+
+                # Extend context with the prediction for next iteration
+                context_series = context_series.concatenate(chunk_pred)
+                remaining_steps -= chunk_size
+
+            # Combine all predictions
+            if len(predictions_list) == 1:
+                predictions = predictions_list[0]
+            else:
+                predictions = predictions_list[0]
+                for pred_chunk in predictions_list[1:]:
+                    predictions = predictions.concatenate(pred_chunk)
+            logger.info(f"Generated {len(predictions)} predictions")
+
+            # Truncate validation series to match prediction length
+            actual = val_series[: len(predictions)]
+
+            # Use original predictions for evaluation (no artificial perturbation)
+            predictions_for_eval = predictions
+
+            # Calculate metric on predictions
+            if metric == "smape":
+                score = smape(actual, predictions_for_eval)
```

Comment:
If both actual and predicted hit zero, denominator becomes zero. Guard against this.

Reply:
done